extensions/transferdetails/service.go

```diff
@@ -0,0 +1,272 @@
+package transferdetails
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/centrifuge/go-centrifuge/documents"
+	"github.com/centrifuge/go-centrifuge/errors"
+	"github.com/centrifuge/go-centrifuge/extensions"
+	"github.com/centrifuge/go-centrifuge/identity"
+	logging "github.com/ipfs/go-log"
+)
+
+// Service defines specific functions for extension funding
+type Service interface {
+	documents.Service
+
+	// DeriveFromUpdatePayload derives Funding from clientUpdatePayload
+	DeriveFromUpdatePayload(ctx context.Context, req UpdateTransferDetailRequest, identifier []byte) (documents.Model, error)
+
+	// DeriveFromPayload derives Funding from clientPayload
+	DeriveFromPayload(ctx context.Context, req CreateTransferDetailRequest, identifier []byte) (documents.Model, error)
+
+	// DeriveFundingResponse returns a funding in client format
+	DeriveTransferResponse(ctx context.Context, model documents.Model, transferID string) (*TransferDetailResponse, error)
+
+	// DeriveFundingListResponse returns a funding list in client format
+	DeriveTransferListResponse(ctx context.Context, model documents.Model) (*TransferDetailListResponse, error)
+}
+
+// service implements Service and handles all funding related persistence and validations
+type service struct {
+	documents.Service
+	tokenRegistry documents.TokenRegistry
+	idSrv         identity.Service
+}
+
+const (
+	transfersLabel    = "transfer_details"
+	transfersFieldKey = "transfer_details[{IDX}]."
+	idxKey            = "{IDX}"
+	transferIDLabel   = "transfer_id"
+)
+
+// DefaultService returns the default implementation of the service.
+func DefaultService(
+	srv documents.Service,
+	tokenRegistry documents.TokenRegistry,
+) Service {
+	return service{
+		Service:       srv,
+		tokenRegistry: tokenRegistry,
+	}
+}
+
+var transfersLog = logging.Logger("tranfers-api")
+
+func deriveDIDs(data *TransferDetailData) ([]identity.DID, error) {
+	var c []identity.DID
+	for _, id := range []string{data.SenderId, data.RecipientId} {
+		if id != "" {
+			did, err := identity.NewDIDFromString(id)
+			if err != nil {
+				return nil, err
+			}
+			c = append(c, did)
+		}
+	}
+
+	return c, nil
+}
+
+func (s service) DeriveFromPayload(ctx context.Context, req CreateTransferDetailRequest, identifier []byte) (documents.Model, error) {
+	model, err := s.GetCurrentVersion(ctx, identifier)
```

Comment:
This method could be more generic.

Where you pass in `transfersFieldKey, transfersLabel`  and the `dids`